### DISCUSSION TITLE
[RFC] New Component: Masonry

### DISCUSSION THREAD

**Discussion by OysterD3** [ant-design/ant-design#51659] [RFCs]
# RFC: Masonry Component

## Motivation
- Create a Pinterest-style grid layout with varying height items but fixed width
- Traditional grid systems often leave undesirable gaps when items have varying heights

## Component API

| Props | Description | Type | Default |
|-------|-------------|------|---------|
| `columns` | Number of columns in the masonry grid layout<br>- Single number for uniform number of columns<br>- Object for responsive number of columns | `number \| { xs?: number; sm?: number; md?: number }` | **Required** |
| `gutter` | Spacing between items. Can be:<br>- Single number for uniform spacing<br>- Object for responsive spacing<br>- Tuple for [horizontal, vertical] spacing | `number \| { xs?: number; sm?: number; md?: number } \| [number, number]` | `0` |
| `sequential` | When true, items are placed sequentially (left-to-right, top-to-bottom). When false, items are placed to minimize vertical gaps | `boolean` | `false` |
| `children` | Child elements to be arranged in the masonry layout | `React.ReactNode` | **Required** |

### Basic Usage
```jsx

function Gallery() {
  return (
    <Masonry columns={3} gutter={16}>
      <img src="image1.jpg" alt="Item 1" />
      <img src="image2.jpg" alt="Item 2" />
      <img src="image3.jpg" alt="Item 3" />
    </Masonry>
  );
}
```

### Responsive Gutters
```jsx

function ResponsiveGallery() {
  return (
    <Masonry 
      columns={4}
      gutter={{
        xs: 8,
        sm: 16,
        md: 24
      }}
    >
      {items.map(item => (
        <Card key={item.id}>
          <img src={item.image} alt={item.title} />
          <h3>{item.title}</h3>
        </Card>
      ))}
    </Masonry>
  );
}
```

### Different Horizontal/Vertical Spacing
```jsx

function GalleryWithCustomSpacing() {
  return (
    <Masonry 
      columns={3}
      gutter={[16, 24]} // [horizontal, vertical]
      sequential={true}
    >
      {items.map(item => (
        <GalleryItem key={item.id} {...item} />
      ))}
    </Masonry>
  );
}
```

## References
- https://mui.com/material-ui/react-masonry/
- https://www.npmjs.com/package/react-responsive-masonry

**Comment by zombieJ** [comment]
`items` 会比较好，然后提供一个 render 方法。`children` 里的元素是不可控的，其实也不支持 HOC：

```tsx
const MyChildren = () => <>{node1}{node2}</>

<Masonry>
  <MyChildren />
</Masonry>
```

**Reply by zombieJ** [reply]
```tsx
type ItemType = {
  height?: number;
  children?: ReactNode;
  key?: ReactKey;
};
```

**Reply by zombieJ** [reply]
key 为 column index 的绑定关系，当提供的时候 key 所在的 column 唯一稳定，如果没有的时候则 items 变化全部重拍
